mypy/typestate.py

```diff
@@ -62,3 +91,99 @@ def record_subtype_cache_entry(cls, left: Instance, right: Instance) -> None:
+        generated here after all modules are type checked anf we have recorded all the subtype
+        checks.
+        """
+        deps = {}  # type: Dict[str, Set[str]]
+        for info in cls._rechecked_types:
+            for attr in cls._checked_against_members[info]:
+                # The need for full MRO here is subtle, during an update, base classes of
+                # a concrete class may not be reprocessed, so not all <B.x> -> <C.x> deps
+                # are added.
+                for base_info in info.mro[:-1]:
+                    trigger = make_trigger('%s.%s' % (base_info.fullname(), attr))
+                    if 'typing' in trigger or 'builtins' in trigger:
+                        # TODO: avoid everything from typeshed
+                        continue
+                    deps.setdefault(trigger, set()).add(make_trigger(info.fullname()))
+            for proto in cls._attempted_protocols[info]:
+                trigger = make_trigger(info.fullname())
+                if 'typing' in trigger or 'builtins' in trigger:
+                    continue
+                # If any class that was checked against a protocol changes,
+                # we need to reset the subtype cache for the protocol.
+                #
+                # Note: strictly speaking, the protocol doesn't need to be
+                # re-checked, we only need to reset the cache, and its uses
+                # elsewhere are still valid (unless invalidated by other deps).
+                deps.setdefault(trigger, set()).add(proto)
+        return deps
+
+    @classmethod
+    def update_protocol_deps(cls, second_map: Optional[Dict[str, Set[str]]] = None) -> None:
+        """Update global protocol dependency map.
+
+        We update the global map incrementally, using a snapshot only from recently
+        type checked types. If second_map is given, update it as well. This is currently used
+        by FineGrainedBuildManager that maintains normal (non-protocol) dependencies.
+        """
+        if cls.proto_deps is None:
+            # Unsuccesful cache loading, nothing to do.
+            return
+        new_deps = cls._snapshot_protocol_deps()
+        for trigger, targets in new_deps.items():
+            cls.proto_deps.setdefault(trigger, set()).update(targets)
+        if second_map is not None:
+            for trigger, targets in new_deps.items():
+                second_map.setdefault(trigger, set()).update(targets)
+        cls._rechecked_types.clear()
+
+    @classmethod
+    def add_all_protocol_deps(cls, deps: Dict[str, Set[str]]) -> None:
+        """Add all known protocol dependencies to deps.
+
+        This is used by tests and debug output, and also when passing
+        all collected or loaded dependencies on to FineGrainedBuildManager
+        in its __init__.
+        """
+        cls.update_protocol_deps()  # just in case
+        if TypeState.proto_deps is None:
```

Comment:
Similar to above, silently failing seems kind of suspicious. I'd rather perform the check in the call site perhaps, and only when it's actually safe to do so.